### PR TITLE
Cleanup for the wikidata plugin

### PR DESCRIPTION
**Pull request by mineo**
These are just a few small commits I made while trying to debug it.

The individual commits have more details where required.

### CODE REVIEW

File: plugins/wikidata/wikidata.py [hunk 1]
```diff
@@ -25,69 +25,65 @@ def __init__(self):
...
-                self.taggers[item_id].append(tagger)
-                self.lock.release()
+            self.process_request(metadata, album, item_id, type='artist')
+
+    def process_request(self, metadata, album, item_id, type):
+        with self.lock:
+            log.debug('WIKIDATA: Looking up cache for item  %s' % item_id)
+            log.debug('WIKIDATA: requests %s' % album._requests)
+            log.debug('WIKIDATA: TYPE %s' % type)
+            if item_id in list(self.cache.keys()):
```
Comment (zas) [inline plugins/wikidata/wikidata.py:51]: You don't need to call `.keys()`, `k in dict` tests again keys.

```python
if item_id in self.cache:
```

File: plugins/wikidata/wikidata.py [hunk 2]
```diff
@@ -25,69 +25,65 @@ def __init__(self):
...
-            path = '/ws/2/%s/%s' % (type, item_id)
-            queryargs = {"inc": "url-rels"}
-            self.ws.get(host, port, path,
-                           partial(self.musicbrainz_release_lookup,
-                                   item_id, metadata),
-                           parse_response_type="xml", priority=False, important=False, queryargs=queryargs)
+            else:
+                # pending requests are handled by adding the metadata object to a
+                # list of things to be updated when the genre is found
+                if item_id in list(self.requests.keys()):
```
Comment (zas) [inline plugins/wikidata/wikidata.py:63]:
```python
if item_id in self.requests:
```

File: plugins/wikidata/wikidata.py [hunk 3]
```diff
@@ -25,69 +25,65 @@ def __init__(self):
...
+                # find the wikidata url if this exists
+                host = config.setting["server_host"]
+                port = config.setting["server_port"]
+
+                path = '/ws/2/%s/%s' % (type, item_id)
+                queryargs = {"inc": "url-rels"}
+                self.ws.get(host, port, path,
+                               partial(self.musicbrainz_release_lookup,
+                                       item_id, metadata),
+                               parse_response_type="xml", priority=False, important=False, queryargs=queryargs)
```
Comment (zas) [inline plugins/wikidata/wikidata.py]: Funky indentation here

File: plugins/wikidata/wikidata.py [hunk 4]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
-        for tagger in self.taggers[item_id]:
-            tagger._requests -= 1
-            if tagger._requests == 0:
-                tagger._finalize_loading(None)
-            log.info('WIKIDATA:  TOTAL REMAINING REQUESTS %s' %
-                     tagger._requests)
-        del self.requests[item_id]
-        self.lock.release()
+        with self.lock:
+            if len(genre_list) > 0:
```
Comment (zas) [inline plugins/wikidata/wikidata.py:168]: No need to call `len()` to test for emptyness:

```python
if genre_list:
```

File: plugins/wikidata/wikidata.py [hunk 5]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
+            else:
+                log.info('WIKIDATA: Genre not found in wikidata')
+
+            log.info('WIKIDATA: Seeing if we can finalize tags %s  ' %
+                     len(self.albums[item_id]))
+
+            for album in self.albums[item_id]:
+                album._requests -= 1
+                album._finalize_loading(None)
+                log.info('WIKIDATA:  TOTAL REMAINING REQUESTS %s' %
```
Comment (zas) [inline plugins/wikidata/wikidata.py:191]: Extra space after `WIKIDATA:`, not sure if intentional or not.

File: plugins/wikidata/wikidata.py [hunk 6]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
     def process_track(self, album, metadata, trackXmlNode, releaseXmlNode):
         self.ws = album.tagger.webservice
         self.log = album.log
-        tagger = album
 
-        item_id = dict.get(metadata, 'musicbrainz_releasegroupid')[0]
-        log.debug('WIKIDATA: looking up release metadata for %s ' % item_id)
-        self.process_request(metadata, tagger, item_id, type='release-group')
+        for release_group in dict.get(metadata, 'musicbrainz_releasegroupid', []):
+            item_id = release_group
```
Comment (zas) [inline plugins/wikidata/wikidata.py]: `release_group` or `item_id` is useless here.

```python
for item_id in dict.get(metadata, 'musicbrainz_releasegroupid', []):
```
Or, few lines below:

```python
self.process_request(metadata, album, release_group, type='release-group')
```

File: plugins/wikidata/wikidata.py [hunk 7]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
+            log.debug('WIKIDATA: looking up release group metadata for %s ' % item_id)
+            self.process_request(metadata, album, item_id, type='release-group')
 
-        for artist in dict.get(metadata, 'musicbrainz_albumartistid'):
+        for artist in dict.get(metadata, 'musicbrainz_albumartistid', []):
             item_id = artist
```
Comment (zas) [inline plugins/wikidata/wikidata.py]: Get rid of useless intermediate variable `item_id` ?

File: plugins/wikidata/wikidata.py [hunk 8]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
             log.info('WIKIDATA: processing release artist %s' % item_id)
-            self.process_request(metadata, tagger, item_id, type='artist')
+            self.process_request(metadata, album, item_id, type='artist')
 
         for artist in dict.get(metadata, 'musicbrainz_artistid'):
             item_id = artist
```
Comment (zas) [inline plugins/wikidata/wikidata.py]: Get rid of useless intermediate variable `item_id` ?

File: plugins/wikidata/wikidata.py [hunk 9]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
             log.info('WIKIDATA: processing track artist %s' % item_id)
-            self.process_request(metadata, tagger, item_id, type='artist')
+            self.process_request(metadata, album, item_id, type='artist')
 
         if 'musicbrainz_workid' in metadata:
             for workid in dict.get(metadata, 'musicbrainz_workid'):
                 item_id = workid
```
Comment (zas) [inline plugins/wikidata/wikidata.py]: Get rid of useless intermediate variable `item_id` ?

File: plugins/wikidata/wikidata.py [hunk 10]
```diff
@@ -25,69 +25,65 @@ def __init__(self):
...
-                tagger._requests += 1
-                self.taggers[item_id].append(tagger)
-                self.lock.release()
+            self.process_request(metadata, album, item_id, type='artist')
+
+    def process_request(self, metadata, album, item_id, type):
+        with self.lock:
+            log.debug('WIKIDATA: Looking up cache for item  %s' % item_id)
+            log.debug('WIKIDATA: requests %s' % album._requests)
+            log.debug('WIKIDATA: TYPE %s' % type)
```
Comment (zas) [inline plugins/wikidata/wikidata.py:50]: `type` is a built-in, i recommend a rename.

File: plugins/wikidata/wikidata.py [hunk 11]
```diff
@@ -25,69 +25,65 @@ def __init__(self):
...
+                if item_id in list(self.requests.keys()):
+                    log.debug(
+                        'WIKIDATA: request already pending, add it to the list of items to update once this has been found')
+                    self.requests[item_id].append(metadata)
+
+                    album._requests += 1
+                    self.albums[item_id].append(album)
+                    return
+                self.requests[item_id] = [metadata]
+                album._requests += 1
```
Comment (zas) [inline plugins/wikidata/wikidata.py:72]: Is this counter really needed ? You may just use `count()` on the list, it is more reliable.

File: plugins/wikidata/wikidata.py [hunk 12]
```diff
@@ -172,60 +163,58 @@ def parse_wikidata_response(self, item, item_id, response, reply, error):
...
-        del self.requests[item_id]
-        self.lock.release()
+        with self.lock:
+            if len(genre_list) > 0:
+                log.info('WiKIDATA: final list of wikidata id found: %s' %
+                         genre_entries)
+                log.info('WIKIDATA: final list of genre: %s' % genre_list)
+
+                log.debug('WIKIDATA: total items to update: %s ' %
+                          len(self.requests[item_id]))
```
Comment (zas) [inline plugins/wikidata/wikidata.py:174]: `len()` returns an integer but the format string uses `%s`

Also, not that `log.debug()` will format arguments:

```python
log.debug('WIKIDATA: total items to update: %d', len(self.requests[item_id]))
```

**Review by zas — COMMENTED** [review]
Sorry for this very late review, i don't usually watch picard-plugins.

**Comment by mineo** [comment]
@zas Good comments, I've fixed the one in code where I did more than change the indentation, but I'm not going to fix everything that's wrong here.

**Comment by mineo** [comment]
@zas Any chance of a re-review?

**Review by zas — APPROVED** [review]

**Pull request merged**